src/sagemaker/remote_function/core/pipeline_variables.py

```diff
@@ -322,39 +328,43 @@ def convert_pipeline_variables_to_pickleable(func_args: Tuple, func_kwargs: Dict
         func_args: function args.
         func_kwargs: function kwargs.
     """
+    converted_func_args = tuple(_convert_pipeline_variable_to_pickleable(arg) for arg in func_args)
+    converted_func_kwargs = {
+        key: _convert_pipeline_variable_to_pickleable(arg) for key, arg in func_kwargs.items()
+    }
 
-    from sagemaker.workflow.entities import PipelineVariable
-
-    from sagemaker.workflow.execution_variables import ExecutionVariables
-
-    from sagemaker.workflow.function_step import DelayedReturn
-
-    # Notes:
-    # 1. The s3_base_uri = s3_root_uri + pipeline_name, but the two may be unknown
-    # when defining function steps. After step-level arg serialization,
-    # it's hard to update the s3_base_uri in pipeline compile time.
-    # Thus set a placeholder: _S3BaseUriIdentifier, and let the runtime job to resolve it.
-    # 2. For saying s3_root_uri is unknown, it's because when defining function steps,
-    # the pipeline's sagemaker_session is not passed in, but the default s3_root_uri
-    # should be retrieved from the pipeline's sagemaker_session.
-    def convert(arg):
-        if isinstance(arg, DelayedReturn):
-            return _DelayedReturn(
-                uri=[
-                    _S3BaseUriIdentifier(),
-                    ExecutionVariables.PIPELINE_EXECUTION_ID._pickleable,
-                    arg._step.name,
-                    "results",
-                ],
-                reference_path=arg._reference_path,
-            )
+    return converted_func_args, converted_func_kwargs
 
-        if isinstance(arg, PipelineVariable):
-            return arg._pickleable
 
-        return arg
+def _convert_pipeline_variable_to_pickleable(arg):
+    """Convert a pipeline variable to pickleable."""
+    from sagemaker.workflow.entities import PipelineVariable
 
-    converted_func_args = tuple(convert(arg) for arg in func_args)
-    converted_func_kwargs = {key: convert(arg) for key, arg in func_kwargs.items()}
+    from sagemaker.workflow.function_step import DelayedReturn
 
-    return converted_func_args, converted_func_kwargs
+    if isinstance(arg, DelayedReturn):
+        # Notes:
+        # 1. The s3_base_uri = s3_root_uri + pipeline_name, but the two may be unknown
+        # when defining function steps. After step-level arg serialization,
+        # it's hard to update the s3_base_uri in pipeline compile time.
+        # Thus set a placeholder: _S3BaseUriIdentifier, and let the runtime job to resolve it.
+        # 2. For saying s3_root_uri is unknown, it's because when defining function steps,
+        # the pipeline's sagemaker_session is not passed in, but the default s3_root_uri
+        # should be retrieved from the pipeline's sagemaker_session.
+
+        container_args = get_step(arg)._properties.AlgorithmSpecification.ContainerArguments
+        execution_id = container_args[11]._pickleable
```

Comment:
I'd recommend using the OutputConfig in the CreateTrainingJob API. 

Also, avoid hardcoding the array item index like `container_args[11]`. Other people might add a new argument and create a bug.

Reply:
Good call out! Updated